fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/metadata/SyntheticRecordTypeBuilder.java

```diff
@@ -0,0 +1,185 @@
+     */
+    @Nonnull
+    public List<C> getConstituents() {
+        return constituents;
+    }
+
+    @Nonnull
+    protected abstract C newConstituent(@Nonnull String name, @Nonnull RecordTypeBuilder recordType);
+
+    /**
+     * Add a new constituent.
+     * @param constituent the new constituent
+     * @return the newly added constituent
+     */
+    @Nonnull
+    protected C addConstituent(@Nonnull C constituent) {
+        constituents.add(constituent);
+        return constituent;
+    }
+
+    /**
+     * Add a new constituent by name.
+     * @param name the correlation name for the new constituent
+     * @param recordType the record type for the new constituent
+     * @return the newly added constituent
+     */
+    @Nonnull
+    public C addConstituent(@Nonnull String name, @Nonnull RecordTypeBuilder recordType) {
+        return addConstituent(newConstituent(name, recordType));
+    }
+
+    /**
+     * Add a new constituent by name.
+     * @param name the correlation name for the new constituent
+     * @param recordType the name of the record type for the new constituent
+     * @return the newly added constituent
+     */
+    @Nonnull
+    public C addConstituent(@Nonnull String name, @Nonnull String recordType) {
+        return addConstituent(name, metaDataBuilder.getRecordType(recordType));
+    }
+
+    /**
+     * Add a new constituent with a correlation name that is the same as the record type name.
+     * @param recordType the record type for the new constituent
+     * @return the newly added constituent
+     */
+    @Nonnull
+    public C addConstituent(@Nonnull RecordTypeBuilder recordType) {
+        return addConstituent(recordType.getName(), recordType);
+    }
+
+    /**
+     * Add a new constituent with a correlation name that is the same as the record type name.
+     * @param constituent name of the record type for the new constituent
+     * @return the newly added constituent
+     */
+    @Nonnull
+    protected C addConstituent(@Nonnull String constituent) {
+        return addConstituent(constituent, constituent);
+    }
+
+    @Nonnull
+    @SuppressWarnings("squid:S1452")
+    public abstract SyntheticRecordType<?> build(@Nonnull RecordMetaData metaData, @Nonnull Descriptors.FileDescriptor fileDescriptor);
+
+    public void buildDescriptor(@Nonnull DescriptorProtos.FileDescriptorProto.Builder fileDescriptorProto) {
+        final DescriptorProtos.DescriptorProto.Builder descriptorProto = fileDescriptorProto.addMessageTypeBuilder();
+        descriptorProto.setName(name);
+        int fieldNumber = 0;
+        for (Constituent constituent : constituents) {
+            descriptorProto.addFieldBuilder()
+                    .setName(constituent.getName())
+                    .setNumber(++fieldNumber)
+                    .setType(DescriptorProtos.FieldDescriptorProto.Type.TYPE_MESSAGE)
+                    .setTypeName(constituent.getRecordType().getDescriptor().getFullName());
```

Comment:
I think, after working on the fix for #508 that this should probably be `"." + ` the full name. It probably doesn't actually matter, but I think it makes it less likely that the type name is set to something that might be ambiguous.

Reply:
Fixed.